### DISCUSSION TITLE
Is there any way to enable multiple threads when adding files?

### DISCUSSION THREAD

**Discussion by FrostyLeaf** [treeverse/dvc#7296] [General]
We found DVC could be insanely slow when adding lots of small files (100k files, each one has roughly 150KiB size) since it calc hash and transfer files with only one thread.

  ```bash
  $ time dvc add images/
  Adding...
Bus error (core dumped)erring                                 67.9k/118k [18:59<14:39,  57.2file/s]
  
  real    41m10.129s
  user	37m54.970s
  sys	1m49.204s
  ```

As you can see, it took more than 40 minutes to calc hash and transfer files, and eventually... failed.

We also tested this dataset with Git LFS, it only took 5m10s to finish adding all of the image files. As Git itself, Git LFS always tends to use multiple threads to handle heavy jobs.

**Comment by daavoo** [comment]
You can set the number of threads for hashing with the option `core.checksum_jobs` of `dvc config` . See https://dvc.org/doc/command-reference/config#core

**Comment by FrostyLeaf** [comment]
I have 4 CPU cores. According to the docs, it should be 2 by default, so there is already more than 1 thread? However, when I checked In top, DVC always took ~ 100% CPU.

**Comment by shcheklein** [comment]
@FrostyLeaf could you please run `dvc version` and share its output please? to give us a bit more information. Have you tried `--job` or `core.checksum_jobs` as @daavoo suggested?

Also would be great if you could run https://github.com/iterative/dvc/wiki/Debugging,-Profiling-and-Benchmarking-DVC on a smaller subset of files to see why is it happening.

**Reply by FrostyLeaf** [reply]
It's 2.9.3. I haven't tried these configurations yet. However, according to the docs, I believe the jobs should be 2 by default since I have 4 CPU cores, right? Any way, I am resetting my whole environment. I'll give u the update asap, I think it will be tomorrow.

**Reply by shcheklein** [reply]
Thanks @FrostyLeaf ! Looking forward to it. And just sharing `dvc version` would already help us :)

**Reply by FrostyLeaf** [reply]
```bash
$ time dvc add -j 4 -vv images/
2022-01-21 22:05:08,245 TRACE: Namespace(cd='.', cmd='add', cprofile=False, cprofile_dump=None, desc=None, external=False, file=None, func=<class 'dvc.command.add.CmdAdd'>, glob=False, instrument=False, instrument_open=False, jobs=4, no_commit=False, out=None, pdb=False, quiet=0, recursive=False, remote=None, targets=['images/'], to_remote=False, verbose=2, version=None, yappi=False)
2022-01-21 22:05:08,398 DEBUG: Adding '/data/test/local/dvc/.dvc/config.local' to gitignore file.
2022-01-21 22:05:08,401 DEBUG: Adding '/data/test/local/dvc/.dvc/tmp' to gitignore file.
2022-01-21 22:05:08,402 DEBUG: Adding '/data/test/local/dvc/.dvc/cache' to gitignore file.
2022-01-21 22:05:08,402 ERROR: --jobs can't be used without --to-remote
------------------------------------------------------------
Traceback (most recent call last):
  File "dvc/command/add.py", line 21, in run
  File "dvc/utils/collections.py", line 161, in inner
  File "dvc/repo/add.py", line 81, in check_arg_combinations
dvc.exceptions.InvalidArgumentError: --jobs can't be used without --to-remote
------------------------------------------------------------
2022-01-21 22:05:08,408 DEBUG: Analytics is enabled.
2022-01-21 22:05:08,439 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmppu40z3te']'
2022-01-21 22:05:08,440 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmppu40z3te']'
[909] Failed to execute script '__main__' due to unhandled exception!
```

So, if I want it to spawn more than one thread, I have to set a remote store? Emmm, I prefer local store for now. There are too many files, a remote one could be make the situation even worse.

**Reply by shcheklein** [reply]
Hey, @FrostyLeaf thanks for trying this. I think though you should use `core.checksum_jobs` in the dvc config, like @daavoo suggested above. Also please share `dvc version` when you have time.

(`-j` specifies number of jobs for upload/download operations)

**Reply by FrostyLeaf** [reply]
```bash
$ dvc version
DVC version: 2.9.3 (rpm)
---------------------------------
Platform: Python 3.8.3 on Linux-5.4.162-86.275.amzn2.x86_64-x86_64-with-glibc2.14
Supports:
	azure (adlfs = 2021.10.0, knack = 0.9.0, azure-identity = 1.7.1),
	gdrive (pydrive2 = 1.10.0),
	gs (gcsfs = 2022.1.0),
	hdfs (fsspec = 2022.1.0, pyarrow = 6.0.1),
	webhdfs (fsspec = 2022.1.0),
	http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	s3 (s3fs = 2022.1.0, boto3 = 1.20.24),
	ssh (sshfs = 2021.11.2),
	oss (ossfs = 2021.8.0),
	webdav (webdav4 = 0.9.3),
	webdavs (webdav4 = 0.9.3)

$ git init
$ dvc init
$ dvc config core.checksum_jobs 4
$ time dvc add -vv images/
...
2022-01-23 01:55:31,355 TRACE: Path '/data/test/local/dvc/images/000000092554.jpg' inode '12724094'
2022-01-23 01:55:31,360 TRACE: Path '/data/test/local/dvc/images/000000150843.jpg' inode '12724095'
2022-01-23 01:55:31,362 TRACE: Path '/data/test/local/dvc/images/000000092554.jpg' inode '12724094'
2022-01-23 01:55:31,363 DEBUG: state save (12724094, 1408149834000000000, 145899) 6d06952eb602aad15596cbf4df923ffe
2022-01-23 01:55:31,363 TRACE: Path '/data/test/local/dvc/images/000000062969.jpg' inode '12724093'
2022-01-23 01:55:31,366 DEBUG: state save (12724093, 1408152868000000000, 240990) dabfd8600ae24d5ddf08c8086bdea862
2022-01-23 01:55:31,363 TRACE: Path '/data/test/local/dvc/images/000000150843.jpg' inode '12724095'
2022-01-23 01:55:31,369 DEBUG: state save (12724095, 1408151556000000000, 82255) 3beeb64a33b6e6cca9b6b0ae02bd3da0
2022-01-23 01:55:31,372 TRACE: Path '/data/test/local/dvc/images/000000574837.jpg' inode '12724096'
Adding...                                                                                                                                                                                                                                    ^2022-01-23 01:55:31,375 TRACE: Path '/data/test/local/dvc/images/000000431854.jpg' inode '12724097'                                                                                                                |10.1k [00:58,    167md5/s]
```

`dvc add -vv` outputs too much, so I have to interrupt it, and clear the workspace and init it again. In the 2nd time to exec `dvc add`, at first the computing speed is definitely faster than before, roughly `300md5/s`. However, this number decreased very soon: 

```bash
$ time dvc add images/
Adding...
Computing file/dir hashes (only done once)                                                 |22.7k [02:05,    128md5/s]
```

**Reply by FrostyLeaf** [reply]
And now it's `73.1md5/s`

**Reply by shcheklein** [reply]
thanks @FrostyLeaf , could you run `dvc version` after `git init` and `dvc init` please.

Also, is it a flat structure - one directory with all 100K images in it, or it has multiple subdirectories in it?

I can try to reproduce this. I think I've seen it once, but the structure was very deep, nested.

**Reply by FrostyLeaf** [reply]
```bash
dvc version
DVC version: 2.9.3 (rpm)
---------------------------------
Platform: Python 3.8.3 on Linux-5.4.162-86.275.amzn2.x86_64-x86_64-with-glibc2.14
Supports:
	azure (adlfs = 2021.10.0, knack = 0.9.0, azure-identity = 1.7.1),
	gdrive (pydrive2 = 1.10.0),
	gs (gcsfs = 2022.1.0),
	hdfs (fsspec = 2022.1.0, pyarrow = 6.0.1),
	webhdfs (fsspec = 2022.1.0),
	http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	s3 (s3fs = 2022.1.0, boto3 = 1.20.24),
	ssh (sshfs = 2021.11.2),
	oss (ossfs = 2021.8.0),
	webdav (webdav4 = 0.9.3),
	webdavs (webdav4 = 0.9.3)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/nvme1n1
Caches: local
Remotes: None
Workspace directory: ext4 on /dev/nvme1n1
Repo: dvc, git
```

It's a flat one.

**Reply by shcheklein** [reply]
thanks!

@FrostyLeaf I see that's it AWS instance, right? what is the exact spec for it and its storage - so that I could launch the same. 

Have you tried other machines by chance?

**Reply by shcheklein** [reply]
I was able to reproduce this on ~ same machine. Performance degrades very fast (it's 60 mds vs 300 mds at 100K files). + I can see that it's utilizing singe thread

cc @daavoo

**Reply by efiop** [reply]
For the record: this is related to a buch of issues that we are working on pre-requsistes for right now. To link a few:

https://github.com/iterative/dvc/issues/5562
https://github.com/iterative/dvc/issues/3177
https://github.com/iterative/dvc/issues/2891

This will be taken a very close look into in https://github.com/iterative/dvc-data that we are currently working on (https://github.com/iterative/dvc/issues/6262)

Also, the metric itself is kinda wrong, as `md5s/sec` is rather meaningless. But once again, this is something that we have on our TODO list for dvc-data in the near future. Please stay tuned. A workaround for now might be to pack the dir into an archive and add that archive into dvc as a singe file, which will speed things up a bunch.

**Reply by FrostyLeaf** [reply]
These issues really help. Thanks for the information, @efiop!

**Reply by FrostyLeaf** [reply]
@shcheklein The spec is t3.xlarge, all of the tests were running on a 200Gi gp3 disk with a base performance(3000 IOPS, 125 MiB/s), also tried a 334Gi gp2 disk. The results were basically the same to me. Hope this do some help to u guys.

**Reply by daavoo** [reply]
> I can see that it's utilizing singe thread

I tested locally and it is using 4 threads by default on my machine (from `max(1, min(4, cpu_count() // 2))`). 

I traced with [viztracer](https://github.com/gaogaotiantian/viztracer):

<details>

<summary>setup.sh</summary>

```bash
rm -rf .dvc
rm -rf foo.dvc
rm -rf foo
dvc init
mkdir foo
for i in {0..10}
do
    echo i > "foo/$(printf "%03d" "$i").txt"
done
```

</details>

<details>

<summary>test_add.py</summary>

```python
from dvc.api import Repo

repo = Repo(".")
repo.add("foo")
```
</details>

Run: `bash setup.sh && viztracer test_add.py --max_stack_depth 10 && vizviewer result.json`

<img width="1697" alt="Captura de pantalla 2022-01-24 a las 20 38 24" src="https://user-images.githubusercontent.com/12677733/150852449-985aca7a-b598-4c8e-9e03-492e4761b70f.png">

**Reply by daavoo** [reply]
Not sure how useful this is but, from the profile, what I can tell is that it appears that `state.get` and `state.save` used inside [get_file_hash](https://github.com/iterative/dvc/blob/main/dvc/data/stage.py#L80) are blocking the threads. Hardcoding the number of threads to 1 gives the same or better performance.

Following the stack, it seems to be due to calls of internal sqlite3 used by our dependency [diskcache](https://github.com/grantjenks/python-diskcache) for storing the [md5s in `State` ](https://github.com/iterative/dvc/blob/822b501069f4553b76534aadf0b2266fd5c26c24/dvc/state.py#L65).

**Reply by shcheklein** [reply]
Good stuff, @daavoo !

Can be related - https://github.com/iterative/dvc/issues/7129 ?

When did we add the diskcache and is there a way to get the current DB schema (including primary keys and other indexes) + is there a way to enable debug logs to see actual queries it's running and their time.

Can we get the DVC version before and after diskcache and compare them?

**Reply by shcheklein** [reply]
(before == the one that explicit indexes and other optimizations. It might be the thing that is very common to ORM systems - unless we know what we do they are far from being optimal to use)

**Reply by efiop** [reply]
The state situation is known, we are not using it efficiently. diskcache was added in 2.0. Guys, please don't waste your time on this, we are aware and will be getting to it as planned.
